exec/totemknet.c

```diff
@@ -1789,58 +1766,88 @@ static int setup_nozzle(void *knet_context)
 	icmap_get_string(NOZZLE_PREFIX, &prefix_str);
-	icmap_get_string(NOZZLE_MACADDR, &macaddr_str);
+	macaddr_res = icmap_get_string(NOZZLE_MACADDR, &macaddr_str);
+	name_res = icmap_get_string(NOZZLE_NAME, &name_str);
 
-	res = icmap_get_string(NOZZLE_NAME, &name_str);
+	/* Is is being removed? */
+	if (name_res == CS_ERR_NOT_EXIST && instance->nozzle_handle) {
+		remove_nozzle_device(instance);
+		free_nozzle(instance);
+		goto out_free;
+	}
 
-	if ((strcmp(name_str, instance->nozzle_name) == 0) &&
-	    (strcmp(ipaddr_str, instance->nozzle_ipaddr) == 0) &&
-	    (strcmp(prefix_str, instance->nozzle_prefix) == 0) &&
-	    (strcmp(macaddr_str, instance->nozzle_macaddr) == 0)) {
-		/* Nothing has changed */
-		knet_log_printf (LOGSYS_LEVEL_DEBUG, "Nozzle device info not changed");
+	if (!name_str) {
+		/* no nozzle */
 		goto out_free;
 	}
 
-	if (name_str && !ipaddr_str) {
+	if (!ipaddr_str) {
 		knet_log_printf (LOGSYS_LEVEL_ERROR, "No IP address supplied for Nozzle device");
 		goto out_free;
 	}
 
-	if (name_str && ipaddr_str && !prefix_str) {
+	if (!prefix_str) {
 		knet_log_printf (LOGSYS_LEVEL_ERROR, "No prefix supplied for Nozzle IP address");
 		goto out_free;
 	}
 
-	/* Is is being removed? */
-	if (res == CS_ERR_NOT_EXIST && instance->nozzle_name) {
-		remove_nozzle_device(instance);
-		free_nozzle(instance);
+	if (macaddr_str && strlen(macaddr_str) != 17) {
+		knet_log_printf (LOGSYS_LEVEL_ERROR, "macaddr for nozzle device is not in the correct format '%s'", macaddr_str);
+		return -1;
+	}
+	if (!macaddr_str) {
+		macaddr_str = (char*)"54:54:01:00:00:00";
+	}
+
+	if (instance->nozzle_name &&
+	    (strcmp(name_str, instance->nozzle_name) == 0) &&
+	    (strcmp(ipaddr_str, instance->nozzle_ipaddr) == 0) &&
+	    (strcmp(prefix_str, instance->nozzle_prefix) == 0) &&
+	    ((macaddr_str == NULL && instance->nozzle_macaddr == NULL) ||
+	     strcmp(macaddr_str, instance->nozzle_macaddr) == 0)) {
+		/* Nothing has changed */
+		knet_log_printf (LOGSYS_LEVEL_DEBUG, "Nozzle device info not changed");
+		goto out_free;
 	}
-	if (res == CS_OK && name_str) {
+
+	/* Add nodeid into MAC address */
+	memcpy(mac, macaddr_str, 12);
+	snprintf(mac+12, sizeof(mac) - 13, "%02x:%02x",
+		 instance->our_nodeid >> 8,
+		 instance->our_nodeid & 0xFF);
+	knet_log_printf (LOGSYS_LEVEL_INFO, "Local nozzle MAC address is %s", mac);
+
+	if (name_res == CS_OK && name_str) {
 		/* Reconfigure */
 		if (instance->nozzle_name) {
+			remove_nozzle_device(instance);
 			free_nozzle(instance);
 		}
 
 		res = create_nozzle_device(knet_context, name_str, ipaddr_str, prefix_str,
-					   macaddr_str);
+					   mac);
 
-		strncpy(instance->nozzle_name, name_str, NOZZLE_MAX_VALUE_LEN);
-		strncpy(instance->nozzle_ipaddr, ipaddr_str, NOZZLE_MAX_VALUE_LEN);
-		strncpy(instance->nozzle_prefix, prefix_str, NOZZLE_MAX_VALUE_LEN);
-		strncpy(instance->nozzle_macaddr, macaddr_str, NOZZLE_MAX_VALUE_LEN);
+		instance->nozzle_name = strdup(name_str);
```

Comment:
Just for formality (and to please coverity), result of strdup should be checked to not being NULL.